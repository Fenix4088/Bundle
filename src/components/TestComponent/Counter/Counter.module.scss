.counter {
  margin-bottom: 100px;
  position: relative;
  width: 400px;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;

  .display {
    margin-bottom: 15px;
    padding: 10px;
    width: 50%;
    background: #bdbef6;
    border-radius: 10px;
    font-size: 42px;
    font-weight: 700;
  }

  .validation-message {
    position: absolute;
    font-size: 18px;
    bottom: -50px;
    color: red;
  }

  .btn-group {
    width: 50%;
    display: flex;
    justify-content: space-between;
    .btn {
      padding: 6px 20px;
      font-size: 18px;
      font-weight: 500;
      border-radius: 6px;
      border: none;
      cursor: pointer;

      &.inc {
        background: #1eda1e;
      }
      &.dec {
        background: #e62929;
      }
      &:disabled {
        background: grey;
        cursor: auto;
      }
    }
  }
}
