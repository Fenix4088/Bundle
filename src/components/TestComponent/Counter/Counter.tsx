import React, { useState } from 'react';
import styles from './Counter.module.scss';

const Counter: React.FC = () => {
  const [count, setCount] = useState<number>(0);

  const isMax = count >= 3;
  const isMin = count === 0;

  const increment = (): void => {
    !isMax && setCount((s) => s + 1);
  };
  const decrement = (): void => {
    !isMin && setCount((s) => s - 1);
  };

  const renderMessage = (isMax: boolean, isMin: boolean) => {
    if (isMax) {
      return (
        <div aria-label="warning-message" className={styles['validation-message']}>
          3 is a max value
        </div>
      );
    } else if (isMin) {
      return (
        <div aria-label="warning-message" className={styles['validation-message']}>
          0 is a min value
        </div>
      );
    } else {
      return null;
    }
  };

  return (
    <div className={styles['counter']}>
      <div aria-label="display" className={styles['display']}>
        {count}
      </div>
      <div className={styles['btn-group']}>
        <button
          type="button"
          name="increment"
          onClick={increment}
          disabled={isMax}
          className={`${styles['btn']} ${styles['inc']}`}
        >
          Inc
        </button>
        <button
          type="button"
          name="decrement"
          onClick={decrement}
          disabled={isMin}
          className={`${styles['btn']} ${styles['dec']}`}
        >
          Dec
        </button>
      </div>
      {renderMessage(isMax, isMin)}
    </div>
  );
};

Counter.displayName = Counter.name;

export { Counter };
