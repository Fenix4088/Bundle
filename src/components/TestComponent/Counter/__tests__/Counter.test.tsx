import { Counter } from '../Counter';

import { render as renderRtl } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export const renderCounter = () => {
  const render = renderRtl(<Counter />);

  return {
    render,
    display: render.getByLabelText(/display/i),
    warning: render.getByLabelText(/warning-message/i),
    incBtn: render.getByRole('button', {
      name: /inc/i,
    }),
    decBtn: render.getByRole('button', {
      name: /dec/i,
    }),
  };
};

describe('TestComponent unit', () => {
  it('should render properly', () => {
    const {
      display,
      warning,
      decBtn,
      incBtn,
      render: { container },
    } = renderCounter();

    expect(display).toBeInTheDocument();
    expect(display).toHaveTextContent('0');
    expect(warning).toBeInTheDocument();
    expect(warning).toHaveTextContent('0 is a min value');
    expect(incBtn).toBeInTheDocument();
    expect(decBtn).toBeInTheDocument();

    expect(container).toMatchSnapshot(``);
  });

  it('should increase counter', async () => {
    const {
      incBtn,
      decBtn,
      warning,
      display,
      render: { queryByLabelText, getByLabelText },
    } = renderCounter();

    expect(decBtn).toBeDisabled();
    expect(warning).toBeInTheDocument();
    expect(warning).toHaveTextContent('0 is a min value');
    expect(display).toHaveTextContent('0');

    await userEvent.click(incBtn);

    expect(display).toHaveTextContent('1');
    expect(queryByLabelText(/warning-message/i)).not.toBeInTheDocument();

    await userEvent.click(incBtn);
    await userEvent.click(incBtn);
    expect(display).toHaveTextContent('3');
    expect(getByLabelText(/warning-message/i)).toHaveTextContent('3 is a max value');
    expect(incBtn).toBeDisabled();
  });
});
