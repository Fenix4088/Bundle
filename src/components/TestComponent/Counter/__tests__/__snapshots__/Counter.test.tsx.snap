// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestComponent unit should render properly 1`] = `
<div>
  <div
    class="counter"
  >
    <div
      aria-label="display"
      class="display"
    >
      0
    </div>
    <div
      class="btn-group"
    >
      <button
        class="btn inc"
        name="increment"
        type="button"
      >
        Inc
      </button>
      <button
        class="btn dec"
        disabled=""
        name="decrement"
        type="button"
      >
        Dec
      </button>
    </div>
    <div
      aria-label="warning-message"
      class="validation-message"
    >
      0 is a min value
    </div>
  </div>
</div>
`;
